### PR TITLE
make remote-ext work with ws and safe RPCs

### PR DESCRIPTION
**Pull request by kianenigma**
switches remote-externalities to use the safe rpcs via ws, instead of unsafe RPCs and http. 

cc @joao-paulo-parity 

polkadot companion: https://github.com/paritytech/polkadot/pull/3096

### CODE REVIEW

File: utils/frame/remote-externalities/src/lib.rs
Comment (niklasad1) [inline utils/frame/remote-externalities/src/lib.rs:149]: derive `Default` doesn't work?

File: utils/frame/remote-externalities/Cargo.toml
Comment (niklasad1) [inline utils/frame/remote-externalities/Cargo.toml:16]: not used anymore?

```suggestion
```

File: utils/frame/remote-externalities/src/lib.rs
Comment (kianenigma) [reply utils/frame/remote-externalities/src/lib.rs:149]: 1. Enums can't derive default, right? 
2. Will hit the same wall as mentioned here:

>  // NOTE: ideally we would use `DefaultNoBound` here, but not worth bringing in frame-support for that.

Essentially, Derive will force us to bound `B: Default` as well, which is redundant.

File: utils/frame/remote-externalities/Cargo.toml
Comment (niklasad1) [inline utils/frame/remote-externalities/Cargo.toml:16]: the ws-client doesn't have any features, it spawns the background task by itself.

File: utils/frame/remote-externalities/src/lib.rs
Comment (niklasad1) [reply utils/frame/remote-externalities/src/lib.rs:149]: > Enums can't derive default, right?

yes, dumb question :)

File: utils/frame/remote-externalities/src/lib.rs
Comment (niklasad1) [inline utils/frame/remote-externalities/src/lib.rs:324]: tested on a synced "public-node"?

File: utils/frame/remote-externalities/src/lib.rs
Comment (niklasad1) [inline utils/frame/remote-externalities/src/lib.rs:284]: Can you explain this "max"  `PAGE` size?

AFAIU, the `state rpc` is using 1000 why is 512 used here?

**Review by niklasad1 — COMMENTED** [review]
Looks overall good, I have some additional annoying questions ^^

File: utils/frame/remote-externalities/src/lib.rs
Comment (kianenigma) [reply utils/frame/remote-externalities/src/lib.rs:324]: It takes a minute or two, but the slowness is mostly a function of how big the state is. My comment is a bit exaggerated and silly tbh, it is slightly slower than the previous http model, and substantially more stable since it is paged.

File: utils/frame/remote-externalities/src/lib.rs
Comment (kianenigma) [reply utils/frame/remote-externalities/src/lib.rs:284]: page size can be any value less than 1000, I randomly chose the biggest power of 2 in that range :D no reason, can also be 1000, can also be configurable, but neither is probably a big deal at this point.

**Review by niklasad1 — APPROVED** [review]

File: utils/frame/remote-externalities/src/lib.rs
Comment (bkchr) [inline utils/frame/remote-externalities/src/lib.rs:302]:
```suggestion
			let page_len = page.len();
			all_keys.extend(page);

			if page_len < PAGE as usize {
```

**Review by bkchr — APPROVED** [review]

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Comment by niklasad1** [comment]
@kianenigma @bkchr 

Can you merge this to the `rocco-v1` branch too after this is merged? 

I have seen some bug reports some users are using `substrate` and `subxt` in the same repo and might end-up with the both the executor features in the `jsonrpsee-http-client` which is "supposed to mutual exclusive" and the pinned version in Cargo.toml, now when the `jsonrpsee-ws-client` is used instead this is not an issue.

FWIW, the `executor features` is planned to be removed when we have updated substrate to tokio1.

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**
